### ISSUE TITLE
check_csrf is broken in 1.6

### ISSUE THREAD

**Issue by hyperknot** [Pylons/pyramid#2294]
I have the following view_config decorator:
```
@view_config(route_name='ajax_map_delete', permission='logged_in', request_method='POST', check_csrf=True)
```

Which, when receiving a correct CSRF token in header `X-CSRF-Token`, breaks the webserver.

```
2016-01-27 03:05:16,211 ERROR [waitress][waitress] Exception when serving /ajax/map_delete/374
Traceback (most recent call last):
  File "/Users/user/.virtualenvs/maphub_web/lib/python2.7/site-packages/waitress/channel.py", line 336, in service
    task.service()
  File "/Users/user/.virtualenvs/maphub_web/lib/python2.7/site-packages/waitress/task.py", line 169, in service
    self.execute()
  File "/Users/user/.virtualenvs/maphub_web/lib/python2.7/site-packages/waitress/task.py", line 388, in execute
    app_iter = self.channel.server.application(env, start_response)
  File "/Users/user/.virtualenvs/maphub_web/lib/python2.7/site-packages/paste/translogger.py", line 69, in __call__
    return self.application(environ, replacement_start_response)
  File "/Users/user/.virtualenvs/maphub_web/lib/python2.7/site-packages/pyramid/router.py", line 223, in __call__
    response = self.invoke_subrequest(request, use_tweens=True)
  File "/Users/user/.virtualenvs/maphub_web/lib/python2.7/site-packages/pyramid/router.py", line 198, in invoke_subrequest
    response = handle_request(request)
  File "/Users/user/.virtualenvs/maphub_web/lib/python2.7/site-packages/pyramid/tweens.py", line 20, in excview_tween
    response = handler(request)
  File "/Users/user/.virtualenvs/maphub_web/lib/python2.7/site-packages/pyramid_tm/__init__.py", line 101, in tm_tween
    reraise(*exc_info)
  File "/Users/user/.virtualenvs/maphub_web/lib/python2.7/site-packages/pyramid_tm/__init__.py", line 83, in tm_tween
    response = handler(request)
  File "/Users/user/.virtualenvs/maphub_web/lib/python2.7/site-packages/pyramid/router.py", line 145, in handle_request
    view_name
  File "/Users/user/.virtualenvs/maphub_web/lib/python2.7/site-packages/pyramid/view.py", line 541, in _call_view
    response = view_callable(context, request)
  File "/Users/user/.virtualenvs/maphub_web/lib/python2.7/site-packages/pyramid/config/views.py", line 328, in attr_view
    return view(context, request)
  File "/Users/user/.virtualenvs/maphub_web/lib/python2.7/site-packages/pyramid/config/views.py", line 299, in predicate_wrapper
    if not predicate(context, request):
  File "/Users/user/.virtualenvs/maphub_web/lib/python2.7/site-packages/pyramid/config/predicates.py", line 256, in __call__
    return self.check_csrf_token(request, val, raises=False)
  File "/Users/user/.virtualenvs/maphub_web/lib/python2.7/site-packages/pyramid/session.py", line 129, in check_csrf_token
    if strings_differ(request.session.get_csrf_token(), supplied_token):
  File "/Users/user/.virtualenvs/maphub_web/lib/python2.7/site-packages/pyramid/util.py", line 278, in strings_differ
    invalid_bits += not compare_digest(left, right)
TypeError: 'unicode' does not have the buffer interface
```

The same code works perfectly in 1.5.8.

Python 2.7.11, OS X 10.9.5

**Comment by mmerickel** [comment]
Hmm, we started using `hmac.compare_digest` when available. It fails when sent 2 values with differing types.

```
>>> compare_digest(u'foo', 'foo')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'unicode' does not have the buffer interface
>>> compare_digest(u'foo', u'foo')
True
```

**Comment by mmerickel** [comment]
It looks like csrf tokens in pyramid are always unicode whereas headers are native strings.

**Comment by stevepiercy** [comment]
@mmerickel I would like your review on PR #2293, because it is sort of related. I almost merged it last night (as well as backport to 1.5), but saw this issue here come up.

**Comment by mmerickel** [comment]
@stevepiercy that PR is unrelated.

**Comment by bertjwregeer** [comment]
HTTP Headers are required to be US-ASCII. ISOwhatsitcalled used to be allowed, but in practice would break things in random weird ways, so ASCII is the only safe charset for HTTP headers.

**Comment by mmerickel** [comment]
@bertjwregeer Yep. It's actually pretty weird that the CSRF token is unicode by default but it's too late to change it now.

**Comment by mmerickel** [comment]
Closed via #2298 and #2299.
